src/airobot/utils/transform_util.py

```diff
@@ -0,0 +1,326 @@
+    pose_final_source = pose_from_matrix(
+        T_pose_final_source, frame_id=pose_source.header.frame_id)
+    return pose_final_source
+
+
+def transform_body(pose_source_world, pose_transform_target_body):
+    """
+    Function to apply a body frame transformation to a 6D pose.
+
+    Args:
+        pose_source_world (PoseStamped): Original source pose to transform 
+        pose_transform_target_body (PoseStamped): Body frame transformation
+            to apply to source pose.
+
+    Returns:
+        PoseStamped: Transformed 6D pose. 
+    """
+    #convert source to target frame
+    pose_source_body = convert_reference_frame(pose_source_world,
+                                               pose_source_world,
+                                               unit_pose(),
+                                               frame_id="body_frame")
+    #perform transformation in body frame
+    pose_source_rotated_body = transform_pose(pose_source_body,
+                                              pose_transform_target_body)
+    # rotate back
+    pose_source_rotated_world = convert_reference_frame(pose_source_rotated_body,
+                                                        unit_pose(),
+                                                        pose_source_world,
+                                                        frame_id="yumi_body")
+    return pose_source_rotated_world
+
+
+def interpolate_pose(pose_initial, pose_final, N):
+    """
+    Function to interpolate between two poses using a combination of
+    linear position interpolation and quaternion spherical-linear
+    interpolation (SLERP)
+
+    Args:
+        pose_initial (PoseStamped): Initial pose
+        pose_final (PoseStamped): Final pose
+        N (int): Number of intermediate points.
+
+    Returns:
+        list: List of poses that interpolates between initial and final pose.
+            Each element is PoseStamped. 
+    """
+    frame_id = pose_initial.header.frame_id
+    pose_initial_list = pose_stamped2list(pose_initial)
+    pose_final_list = pose_stamped2list(pose_final)
+    trans_initial = pose_initial_list[:3]
+    quat_initial = pose_initial_list[3:]
+
+    trans_final = pose_final_list[:3]
+    quat_final = pose_final_list[3:]
+
+    trans_interp_total = [np.linspace(trans_initial[0], trans_final[0], num=N),
+                          np.linspace(trans_initial[1], trans_final[1], num=N),
+                          np.linspace(trans_initial[2], trans_final[2], num=N)]
+    
+    key_rots = R.from_quat([quat_initial, quat_final])
+    slerp = Slerp(np.arange(2), key_rots)
+    interp_rots = slerp(np.linspace(0, 1, N))
+    quat_interp_total = interp_rots.as_quat()    
+
+    pose_interp = []
+    for counter in range(N):
+        pose_tmp = [
+            trans_interp_total[0][counter],
+            trans_interp_total[1][counter],
+            trans_interp_total[2][counter],
+            quat_interp_total[counter][0], #return in ROS ordering w,x,y,z
```

Comment:
is the comment here outdated? seems like the ordering is x,y,z,w

Reply:
fixed in 2259946